Add Scale to Score and change tuning only during playback.

sraduvictor opened the pull request.

sraduvictor commented:
This PR is based on https://github.com/musescore/MuseScore/pull/2404

lasconic commented:
I closed the other PR since this one will be better. I created a discussion here https://musescore.org/en/node/99666

sraduvictor commented:
I changed the diff to support reading scala files with 7 (It was actually 7 not 8 :) ), 12, 17, 21 and 35 notes.
For 35 notes the format I'm assuming (which might be wrong) is :
  D_BB, C_S, D_B, C_SS, D, E_BB, D_S, E_B, D_SS, E, .... , B, C_BB, B_S, C_B, B_SS, C

sraduvictor commented:
Ideas for next diffs : 
  - show the original notes read from scala somewhere in the score panel (lasconic's idea)
  - save the original notes from scala in the musescore file

mgavioli commented on libmscore/scale.h:
```diff
@@ -0,0 +1,50 @@
+
+      std::vector<double> originalNotes;
+      std::map<Tpc, double> computedNotes;
+
```
I see your approach, but it seems to me somewhat more complex than necessary. I believe you only need a fixed array of 35 `double`'s and, when needed, index it with the note TPC1.

sraduvictor replied:
I think both approaches are good. I choose this one because of the following reasons : 
  - it's easier to see which enum I'm using.
  - the code would not look as good trying to write TPC_C - TPC_MIN because TPC_MIN is not 0 but -1.
  - all values are placed in the vector in a manner easy to understand just by looking at scale.cpp, while storing by the index of the enum it would be hard to figure out which note it is unless you go and check that enum (because it starts with all BB and then B, normal, S, and SS).

mgavioli replied:
Also, if I am not mistaken, do we really need fractionary cents? Wouldn't rounding to the nearest `int` enough? I doubt anyone can notice a difference of 0.5 cent.

sraduvictor replied:
I guess that's true. But the difference is so small that I don't mind storing double.

Would you mind adding your editor to this class, so we can change it? (we should change only the original notes). Or, you could send me the code and I'll integrate it.

mgavioli replied:
"_the difference is so small that I don't mind storing double._"

The difference is small memory-wise; less small CPU-wise... Not to mention the 'illusory precision' effect: it is far from certain that the final audible result would be precise even to the cent (with the usual sampling rates around 44Kz, an A at 880Hz only has 50 samples to work with, assuming the synthesizer does not add its own curses), to make worth storing, computing, displaying _n_ decimals.

The original plug-in for ver. 1.x can be found at https://musescore.org/en/project/scales

Please note, though, that I now have better ideas about the features the editor should have to be really useful.

mgavioli commented on libmscore/scale.cpp:
```diff
@@ -0,0 +1,265 @@
+            }
+      return true;
+      }
+
```
I do not see the need for all those hard coded methods. What are needed are the the first and the last TPC 's directly retrievable from the Scala file and four arrays with the destination of each value for each type of scale.

The remaining TPC's can be filled by duplicating the corresponding enharmony, which is 12 steps above for the lower uncovered part or 12 steps below for the upper part.

The 7-note case (or any case which does not cover at least 12 notes) is special indeed, as the missing 'black keys' needs to be computed, but the whole can still be efficiently coded in a single method. This should make easier future maintenance.

sraduvictor replied:
I think your suggestion is for having the keys inputted in the order of the TPC enum, instead of the Scala file format, which I did not implement. I still want to keep the Scala format. If you find there is a case for having a sharp note higher than the next bemol, you could always change the scala file and import it.

mgavioli commented on libmscore/scale.h:
```diff
@@ -0,0 +1,50 @@
+private:
+      QString getNextLine(QTextStream& in);
+
+      std::vector<double> originalNotes;
```
What the `originalNotes` is for? It is never accessed, once `computedNotes` is filled. Some future development?

sraduvictor replied:
mgavioli, would you mind getting on irc to talk about this diff?

I think your suggestion is for having the keys inputted in the order of the TPC enum, instead of the Scala file format, which I did not implement. I still think I'd like to keep the Scala format. If you find there is a case for having a sharp note higher than the next bemol, you could always change the scala file and input it.

originalNotes will be used for showing the scale and saving it to the file.

mgavioli replied:
"_If you find there is a case for having a sharp note higher than the next bemol, you could always change the scala file and input it._".

This would no longer be a valid Scala file. I don't understand why we should force a defined format beyond its intended usage, when a more adequate one (perfectly adequate, I would say) is easy (easier, I would say) to define and manage.

Scala format should be kept ad an import option, given the amount of data available under it. But I maintain it is not adequate as internal data structure.

mgavioli replied:
"_originalNotes will be used for showing the scale and saving it to the file._".

I do not understand: the values are exactly the same as `computedNotes`.

I would understand if the original strings were kept for ratios ("5/4" is more informative than 386.5 cents), but as the values are the same, why storing them twice?

mgavioli commented on libmscore/scale.cpp:
```diff
@@ -0,0 +1,265 @@
+                        return false;
+                        }
+
+                  originalNotes.push_back(1200 * (values[0].toDouble() / values[1].toDouble()) - 1200);
```
Not sure this is correct. If the value is expressed as a ratio, it is a linear value, not proportional to any cent value, which are logarithmic.

What you need is something like this:

`value = log(values[0].toDouble() / values[1].toDouble()) / log(pow(2, 1/1200));`

where `log(pow(2, 1/1200))` is the natural logarithm of 1 cent and can be pre-computed once for all.

sraduvictor replied:
Thanks for the explanation. When looking in http://www.huygens-fokker.org/scala/scl_format.html the ratio value was poorly explained and did not mention anything about logarithms.

mgavioli commented on libmscore/rendermidi.cpp:
```diff
@@ -154,13 +155,21 @@ void Score::updateChannel()
+      if (fabs(scaleTuning) > 200) {
+            int step = (int)scaleTuning / 100;
+            pitch += step;
+            scaleTuning += step * 100;
```
Shouldn't be `scaleTuning -= step * 100;` ?

For instance, if `scaleTuning` is 250, `step` is increased by 2 (= 200 cent) and `scaleTuning` should be correspondingly be reduced  by 200.

mgavioli replied:
Also, I know to be almost alone in this, but I find literal numeric values reader-hostile. I would replace `100` with a symbolic value, say `centPerSemitone`, defined at the top and `200` with `2 * centPreSemitone`.

But, I know, these out-of-fashion concerns, nowadays...

sraduvictor replied:
Thanks for finding this bug.

sraduvictor commented:
I updated the diff, to fix the bug mgavioli found in playNote and to use centPerSemitone instead of 100.

sraduvictor commented:
Updated the diff to use log scale for ratios instead of normal scale.

mgavioli commented on libmscore/scale.cpp:
```diff
@@ -0,0 +1,265 @@
+
+      return standardNote->second - computedNote->second;
+      }
+}
```
One of the advantages of having the scale notes in TPC order is that this whole method -- which is executed once for **each** score note, **each** time it is played -- becomes simply:

`return computedNotes[note->tpc1() - TPC_MIN];`

and it is likely candidate for inlining. Not even need to check the array bounds, as the TPC **must** be valid.

sraduvictor replied:
It might look simpler but the map/vector needs to keep the standardNotes. We also need to check for TPC_INVALID or any new tpc values that are not inserted in the vector. The code would look something like :

  `  <the standard notes vector>
     if (note->tpc1() < TPC_MIN || note->tpc1() - TPC_MIN > 35)
          return 0;
    return computedNotes[note->tpc1() - TPC_MIN] - standardNotes[note->tpc1() - TPC_MIN]`

mgavioli replied:
Oh yes, I see! Currently `computedNotes` contains the note position within the octave, rather than the delta from 12 EDO. But this computation only depends on the TPC and nothing else, it can be done once for all while filling the array, rather than each time a value is accessed.

And `standardNotes` also can simply be an array in TPC order, to be used while filling `computedNotes`.

In fact, you do not even need the `standardNotes` array, as the amount to subtract is:

`(TPC_FIRST_STEP + (tpc - TPC_MIN) * TPC_STEPS_IN_FIFTH) % CENT_PER_OCTAVE`

where
- `TPC_FIRST_STEP` is the octave step of the first TPC (Fbb ~= Eb = 3rd step of an octave; this is already defined in `pitchspelling.h` but commented out);
- `TPC_STEPS_IN_FIFTH` is the number of steps in a fifth = 7;
- `CENT_PER_OCTAVE` = 1200;

I have posted at http://pastebin.com/6sGctEm4 a sample of how the `loadScale()` file could become with this approach: 85 lines of code are enough for the whole Scala import into a self-contained regular C++ array.

About checking the array bounds: a lot of things would go awry, if the tpc of a note is not valid, well before coming to playback. But I agree that at least an `assert` would be necessary to catch bugs.

sraduvictor replied:
Your approach is more compact, but it is also harder to understand and thus maintain or modify.
Also what happens when we add new simbols that won't work with the logic, are we going to rewrite it completely?
Some issues I found : 
 - for TPC_FIRST_STEP = 3, TPC_STEPS_IN_FIFTH = 7, CENT_PER_OCTAVE = 1200
  tpc = TPC_F_BB = TPC_MIN + 1 the result would be :  (3 + 1 * 7) % 1200 = 10, but the result should be 400.
  - you're not support the 17 note variant
  - you should use 14 instead of TPC_DELTA_ENHARMONIC

mgavioli replied:
Less code means less code to understand, less things which could go wrong and so on. That code proceeds by logical steps and it is not only more compact: it supports any (present or future) sequence of TPC, not only the 5 currently implemented, with the only requirement that the sequence shall not have 'holes'. It is necessary to understand it once for all, not each time a new sequence is added. If you don't like the 'big' formula to convert octave offset to 12EDO delta, see below.

I do not understand your comment about new symbols: new TPC's (for instance in the unlikely case MuseScore is going to support triple accidentals) would fall in the logic. As well as for new types of scales. In any case, it would simply be a matter of adding more elements to the `tpcs[][]` array and/or to the `supportedNbs` map at the top.

About your points:
- Correct: I left out a `* centsPerSemitone` factor while copying my notes; as I noted, this is not production code, but something I wrote mostly by memory to show the coding patterns. Note also that for the TPC_F_BB case, the expected value is 300, not 400 and that `tpc - TPC_MIN` equals 0, not 1. So, for this case the reckoning comes out as `((3 + 0*7) * 100) % 1200 = 300`, as expected.

As an aside: if a lookup table (`static const int equalTempCent[TPC_NUM_OF];`) filled with the 12 EDO cents in TPC order is preferred to this formula, ok as well: it would probably spare some CPU cycle. I made up the formula mainly to show that, by using TPC's, everything falls 'automagically' in its place and almost anything can be computed with general rules, rather than with _ad hoc_ cases.
- I **am** supporting the 17 note case: it is the third element of the array at the top (your code is filtering it out, while checking the number of notes at the top, though).
- Can you explain why TPC_F_BB, TPC_C_BB, TPC_E_SS, TPC_B_SS are not set correctly?
- No need to check double accidentals in particular: they fall in the general logic. Anything below `tpcMin` or above `tpcMax` is for sure not set yet.
- Correct: read `for (i = tpcMin-1; i <= TPC_E_B; i--)` instead. Of course this does not set TPC_F_B and TPC_C_B: they do not fall within the minimum 12-note range!
- Symmetrically to the above.

sraduvictor replied:
I thought TPC_F_BB should equal TPC_E so it should be 400. Is this not the case?

mgavioli commented on libmscore/scale.cpp:
```diff
@@ -0,0 +1,265 @@
+bool Scale::loadScale(const QString& fn)
+      {
+      static set<int> supportedNbs = { 8, 12, 21, 35 };
+
```
8 or 7? See also below at line 80.

sraduvictor replied:
It should be 7. Thanks for finding the bug. It was working for me because I was adding 1 to the number of notes.

mgavioli commented on libmscore/scale.cpp:
```diff
@@ -0,0 +1,265 @@
+      computedNotes.clear();
+      switch (originalNotes.size()) {
+            case 7:
+                  computeNotes7();
```
7 or 8? Se also above line 43

sraduvictor replied:
should be 7

mgavioli commented on libmscore/scale.cpp:
```diff
@@ -0,0 +1,265 @@
+            TPC_G_S, TPC_A,
+            TPC_A_S, TPC_B,
+                     TPC_C};
+
```
The most common attribution of black keys is: C#, Eb, F#, G# (occasionally Ab), Bb

sraduvictor replied:
I'll change A# to Bb

mgavioli commented on libmscore/scale.cpp:
```diff
@@ -0,0 +1,265 @@
+      static vector<Tpc> notesToFill = {
+            TPC_C_S, TPC_D_B, TPC_D,
+            TPC_D_S, TPC_E_B, TPC_E,
+            TPC_E_S, TPC_F_B, TPC_F,
```
Not sure. I would expect Fb (usually near to E) to come before E# (usually near to F).

sraduvictor replied:
so should I assume all bemol notes are below sharp notes?

mgavioli replied:
Absolutely **not**!

Fb comes before E# and Cb before B# because between E and F and between B and C there is only one semitone. In all other cases, where between the two notes there is one tone, the lower note sharpened come before the upper note flattened more often than not.

At least according to my experience, which is limited to Western music and (some of!) its scales and temperaments. No idea about other cultures or less well-known scales/temperaments.

mgavioli commented on libmscore/scale.cpp:
```diff
@@ -0,0 +1,265 @@
+            TPC_F_S, TPC_G_B, TPC_G,
+            TPC_G_S, TPC_A_B, TPC_A,
+            TPC_A_S, TPC_B_B, TPC_B,
+            TPC_B_S, TPC_C_B, TPC_C};
```
Same reason as above: Cb, B#, C?

sraduvictor replied:
should I do this for all notes?

mgavioli commented on libmscore/scale.cpp:
```diff
@@ -0,0 +1,265 @@
+      static vector<Tpc> notesToFill = {
+            TPC_D_BB, TPC_C_S, TPC_D_B, TPC_C_SS, TPC_D,
+            TPC_E_BB, TPC_D_S, TPC_E_B, TPC_D_SS, TPC_E,
+            TPC_F_BB, TPC_E_S, TPC_F_B, TPC_E_SS, TPC_F,
```
Also not sure. In fact, I would expect Fbb to come before E, but I finally have no idea if it usually comes before or after Eb. Possibly after, possibly it depends heavily from the individual scale.

I am also not sure if this eventually matters at all: does Scala support 35 notes per octave? I am not sure.

sraduvictor replied:
There are scala files that have 35 notes per octave (there actually are some with a lot more notes).

sraduvictor commented:
mgavioli, first of all let me thank you, for taking the time to review my diff, and for finding so many bugs, and giving me many great suggestions.

I'll change originalNotes to store the originalString so that it is unaltered from the Scala file.
I'll change the order to have bemol notes first and shard notes after.
I could change computedNotes to have the tpc order, but I'd still keep it as a map, so I don't know if that would improve the code by a lot.

If you have any other ideas for improvement please put them here.

I'll also take a look at the plugin and try to think of a way to create an editor. If you have any other ideas please let me know about how it should look like.

sraduvictor commented:
I made the changes I mentioned. I realized computedNotes is a map so it does not have any order, so I didn't know what to change for it.

lasconic commented on libmscore/rendermidi.cpp:
```diff
@@ -154,13 +155,23 @@ void Score::updateChannel()
       velo = note->customizeVelocity(velo);
+
+      double scaleTuning = s.getTuning(note);
+      if (fabs(scaleTuning) > 2 * centsPerSemitone) {
```
This is worth a comment, to explain what we are doing here.

sraduvictor replied:
Good point

sraduvictor commented:
mgavioli, I made another diff with the version you suggested : https://github.com/musescore/MuseScore/pull/2407

Please, take a look at that one, and tell me how I should change it.

sraduvictor commented:
After writing https://github.com/musescore/MuseScore/pull/2407 I realized there were some errors in this diff so I fixed them (the tuning was working the other way, and standardNotes were not set properly).
I also incorporated some other ideas from the other diff.
